lib/elixir/lib/map.ex

```diff
@@ -93,11 +96,10 @@ defmodule Map do
   """
   @spec new(Enum.t, (term -> {key, value})) :: map
   def new(enumerable, transform) do
-    fun = fn el, acc ->
-      {k, v} = transform.(el)
-      put(acc, k, v)
-    end
-    Enum.reduce(enumerable, %{}, fun)
+    enumerable
+    |> Enum.reduce([], &[transform.(&1) | &2])
+    |> :lists.reverse
+    |> :maps.from_list
```

Comment:
Are you suggesting a fast track for Enum.into/2 and maps? The regular collectable implementation for maps can't benefit from this, because it only has one item at a time. `:maps.from_list` only works when you know all the items you want to put into the map ahead of time.